packages/react-reconciler/src/ReactFiberIncompleteWork.js

```diff
@@ -0,0 +1,380 @@
+/**
+ * Copyright (c) 2013-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {HostContext} from './ReactFiberHostContext';
+import type {Fiber} from './ReactFiber';
+import type {FiberRoot} from './ReactFiberRoot';
+import type {CapturedValue} from './ReactCapturedValue';
+import type {UpdateQueue} from './ReactFiberUpdateQueue';
+import type {ExpirationTime} from './ReactFiberExpirationTime';
+
+import {createWorkInProgress} from './ReactFiber';
+import {createCapturedValue} from './ReactCapturedValue';
+import {
+  ensureUpdateQueues,
+  insertUpdateIntoFiber,
+} from './ReactFiberUpdateQueue';
+
+import {
+  IndeterminateComponent,
+  FunctionalComponent,
+  ClassComponent,
+  HostRoot,
+  HostComponent,
+  HostText,
+  HostPortal,
+  CallComponent,
+  CallHandlerPhase,
+  ReturnComponent,
+  ContextProvider,
+  ContextConsumer,
+  Fragment,
+  Mode,
+} from 'shared/ReactTypeOfWork';
+import {
+  NoEffect,
+  DidCapture,
+  Incomplete,
+  Combined,
+  Err,
+  AlgebraicEffectMask,
+} from 'shared/ReactTypeOfSideEffect';
+
+import {
+  popContextProvider as popLegacyContextProvider,
+  popTopLevelContextObject as popTopLevelLegacyContextObject,
+} from './ReactFiberContext';
+import {popProvider} from './ReactFiberNewContext';
+
+import invariant from 'fbjs/lib/invariant';
+
+function captureEffect(effectList, previousEffect, effect) {
+  // Remove an effect from the effect list
+  if (previousEffect === null) {
+    effectList.firstEffect = effect.nextEffect;
+  } else {
+    previousEffect.nextEffect = effect.nextEffect;
+  }
+  if (effect.nextEffect === null) {
+    effectList.lastEffect = previousEffect;
+  }
+}
+
+function addEffectToReturnFiber(returnFiber, effectFiber) {
+  if ((returnFiber.effectTag & Incomplete) === NoEffect) {
+    // Some children already completed. Reset the effect list.
+    returnFiber.firstEffect = returnFiber.lastEffect = effectFiber;
+  } else {
+    // We already have some thrown effects. Append the new effect to the
+    // end of the list.
+    if (returnFiber.lastEffect !== null) {
+      returnFiber.lastEffect.nextEffect = effectFiber;
+    } else {
+      returnFiber.firstEffect = effectFiber;
+    }
+    returnFiber.lastEffect = effectFiber;
+  }
+  returnFiber.effectTag |= Incomplete;
+}
+
+export function raiseUnknownEffect(
+  returnFiber: Fiber,
+  sourceFiber: Fiber,
+  value: mixed,
+) {
+  let algebraicEffectTag;
+  if (value instanceof Error) {
+    algebraicEffectTag = Err;
+  } else if (value !== null && typeof value === 'object') {
+    // If instanceof fails, fall back to duck typing.
```

Comment:
This feels surprising. Moving code into an iframe could fundamentally change the behavior if it used to pass the first check but doesn't pass the second one now.

Reply:
Maybe we should only do duck typing and skip `instanceof` entirely?

Reply:
This would make sense to me. I would maybe check just `stack` since it's added by the browser, but `message` could potentially be empty (I think?)

On the other hand is it plausible that `Promise` ever gets a `stack`?

Also, what happens if you just `throw` a string? It's bad but it can happen. We used to handle this, do we not anymore?

My impression is that the plan is to treat thenables specially, and treat everything else as an error. Why is the logic written around special casing the error object?

Reply:
I want to leave the door open for supporting custom algebraic effects. Error boundaries are a normal effect boundary that pattern matches on errors; there’s another boundary that pattern matches on promises, and so on. That’s what these lines here are intended to model.

Error boundaries will no longer catch null or undefined, but the root still will (that’s why the Unknown type exists).

Reply:
Do error boundaries still catch strings?

Reply:
No they only catch errors. A string will propagate to the root. I’ll add this to the list of known deviations.

If this is too drastic a change I’m fine with for holding off on that part until React 17. Though if we do that, we should warn so people don’t rely on that behavior.

Reply:
To be honest I'd say it's okay to say this was unsupported. It's a known bad pattern, and error boundaries should not (yet :-) be used for control flow anyway. We always suggested they're meant to be used for exceptional situations.

Reply:
I don't think this is quite ok for a breaking change. The cross-realm issue and the fact that throwing strings, while not good practice, is still common.

Thenables are a lot more restrictive since they will essentially always unwrap through any async boundary, they're pretty useless for anything other than as promises. So limiting this to thenable would be a lot safer.

I understand that our general strategy is to attempt to limit componentDidCatch to errors only but not quite sure about the mechanism here.